src/communication/request_sync_client_server.hpp

```diff
@@ -29,15 +29,55 @@ namespace cubcomm
 {
   /* wrapper/helper encapsulating instances of request_client_server, request_sync_send_queue
    * and request_queue_autosend working together
+   *
+   * request_sync_client_server extends the request handling with the ability of sending requests awaiting responses
+   * and the ability to respond.
+   *
+   * the request payloads are preceded by a response sequence number. the request sequence number is generated, packed
+   * and unpacked by request_sync_client_server; its value is not transparent to the users making the requests and the
+     responses. see nested class internal_payload.
    */
   template <typename T_OUTGOING_MSG_ID, typename T_INCOMING_MSG_ID, typename T_PAYLOAD>
   class request_sync_client_server
   {
+    private:
+
+      // Wrap the user payload with a response sequence number
+      using response_sequence_number_t = std::uint64_t;
+      static constexpr response_sequence_number_t NO_RESPONSE =
+	      std::numeric_limits<response_sequence_number_t>::max ();
+
     public:
       using outgoing_msg_id_t = T_OUTGOING_MSG_ID;
-
       using request_client_server_t = cubcomm::request_client_server<T_OUTGOING_MSG_ID, T_INCOMING_MSG_ID>;
-      using incoming_request_handler_t = typename request_client_server_t::server_request_handler;
+
+      // The user payload (of type T_PAYLOAD) is accompanied by a response sequence number set by
+      // request_sync_client_server.
+      class internal_payload : public cubpacking::packable_object
+      {
+	public:
+	  internal_payload () = default;
+	  internal_payload (response_sequence_number_t a_rsn, T_PAYLOAD &&a_payload);
+	  internal_payload (internal_payload &&other);
+	  internal_payload (const internal_payload &other) = delete;
+	  ~internal_payload () = default;
+
+	  internal_payload &operator= (internal_payload &&other);
+	  internal_payload &operator= (const internal_payload &) = delete;
+
+	  void push_payload (T_PAYLOAD &&a_payload);
+	  T_PAYLOAD pull_payload ();
+
+	  size_t get_packed_size (cubpacking::packer &serializator, std::size_t start_offset = 0) const final override;
+	  void pack (cubpacking::packer &serializator) const final override;
+	  void unpack (cubpacking::unpacker &deserializator) final override;
+
+	private:
+	  response_sequence_number_t m_rsn = NO_RESPONSE;
+	  T_PAYLOAD m_user_payload;
+      };
```

Comment:
Somehow this feels a bit anachronic. On one side it is an "internal_payload". On the other side the client still has to deal with it.
Why not unify "internal_payload" here with `request_sync_send_queue::queue_item_type`, integrate it full stack and expose to the client, with the only drawback that the client has to ignore the sequence number but still has to populate the id and the net-payload, while the infrastructure populates the sequence number.
I do not have a better understanding as of now, so it might evolve over time when we get a better hold of the design & usage.